src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java

```diff
@@ -0,0 +1,203 @@
+ * An april tag is like a 2D barcode used to find the robot's position on the field.
+ * Since the tag's position on the field is known, we can calculate our position relative to it, therefore estimating our position on the field.
+ */
+public class AprilTagCamera {
+    protected final String name;
+    private final RobotPoseSourceInputsAutoLogged inputs = new RobotPoseSourceInputsAutoLogged();
+    private final Transform3d robotCenterToCamera;
+    private final AprilTagCameraIO aprilTagCameraIO;
+    private double lastUpdatedTimestamp;
+    private final double solvePNPTranslationsStandardDeviationsExponent, assumedRobotPoseTranslationsStandardDeviationsExponent, solvePNPThetaStandardDeviationsExponent;
+    private Pose2d robotPose = null;
+
+    /**
+     * Constructs a new AprilTagCamera.
+     *
+     * @param robotPoseSourceType                                       the type of camera
+     * @param name                                                      the camera's name
+     * @param robotCenterToCamera                                       the transform of the robot's origin point to the camera
+     * @param solvePNPTranslationsStandardDeviationsExponent            the calibrated gain to calculate the translation deviation from the estimated pose when using solve PNP
+     * @param solvePNPThetaStandardDeviationsExponent                   the calibrated gain to calculate the theta deviation from the estimated pose when using solve PNP
+     * @param assumedRobotHeadingTranslationsStandardDeviationsExponent the calibrated gain to calculate the translation deviation from the estimated pose when getting the pose by assuming the robot's heading
+     */
+    public AprilTagCamera(AprilTagCameraConstants.RobotPoseSourceType robotPoseSourceType, String name, Transform3d robotCenterToCamera, double solvePNPTranslationsStandardDeviationsExponent, double solvePNPThetaStandardDeviationsExponent, double assumedRobotHeadingTranslationsStandardDeviationsExponent) {
+        this.name = name;
+        this.robotCenterToCamera = robotCenterToCamera;
+        this.solvePNPTranslationsStandardDeviationsExponent = solvePNPTranslationsStandardDeviationsExponent;
+        this.solvePNPThetaStandardDeviationsExponent = solvePNPThetaStandardDeviationsExponent;
+        this.assumedRobotPoseTranslationsStandardDeviationsExponent = assumedRobotHeadingTranslationsStandardDeviationsExponent;
+
+        if (Robot.IS_REAL)
+            aprilTagCameraIO = robotPoseSourceType.createIOFunction.apply(name);
+        else
+            aprilTagCameraIO = new AprilTagCameraIO();
+    }
+
+    public void update() {
+        aprilTagCameraIO.updateInputs(inputs);
+        Logger.processInputs("Cameras/" + name, inputs);
+        logVisibleTags();
+
+        robotPose = calculateBestRobotPose(inputs.distanceFromBestTag);
+        logEstimatedRobotPose();
```

Comment:
I think we this should be the order:
update % procces inputs

Calculate robot pose

Log stuff